ament_cmake_python/cmake/ament_python_install_package.cmake

```diff
@@ -116,51 +111,105 @@ setup(
+    # determine the Python package's source dependencies for proper build
+    # invalidation.
+    install(CODE
+      "set(install_dir ${install_dir})
+       set(extra_install_args ${extra_install_args})
+       if(DEFINED ENV{DESTDIR} AND NOT \"\$ENV{DESTDIR}\" STREQUAL \"\")
+         list(APPEND extra_install_args --root \$ENV{DESTDIR})
+         file(TO_CMAKE_PATH \"\$ENV{DESTDIR}/\${install_dir}\" install_dir)
+       endif()
+       message(STATUS
+         \"Installing: ${package_name} as flat Python egg under \${install_dir}\")
+       execute_process(
+         COMMAND
+           \"${PYTHON_EXECUTABLE}\" setup.py install
+             --single-version-externally-managed
+             --install-scripts ${install_scripts_dir}
+             --prefix ${install_prefix}
+             --record install.log
+             \${extra_install_args}
+         WORKING_DIRECTORY \"${build_dir}\"
+         OUTPUT_QUIET
+       )"
+    )
   else()
-    set(extra_install_args --no-compile)
-  endif()
+    # NOTE(hidmic): Make sure build directory makes it into the PYTHONPATH.
+    # See https://github.com/colcon/colcon-core/pull/163 for further reference.
+    file(TO_NATIVE_PATH "${build_dir}" build_prefix)
+
+    if(NOT "${package_name}" STREQUAL "${PROJECT_NAME}")
+      set(hook_name "${package_name}_python_develop")
+    else()
+      set(hook_name "python_develop")
+    endif()
 
-  # Install as flat Python .egg to mimic https://github.com/colcon/colcon-core
-  # handling of pure Python packages.
-
-  # NOTE(hidmic): Allow setup.py install to build, as there is no way to
-  # determine the Python package's source dependencies for proper build
-  # invalidation.
-  install(CODE
-    "set(extra_install_args ${extra_install_args})
-     set(install_dir \"${CMAKE_INSTALL_PREFIX}/${PYTHON_INSTALL_DIR}\")
-     if(DEFINED ENV{DESTDIR} AND NOT \"\$ENV{DESTDIR}\" STREQUAL \"\")
-       list(APPEND extra_install_args --root \$ENV{DESTDIR})
-       file(TO_CMAKE_PATH \"\$ENV{DESTDIR}/\${install_dir}\" install_dir)
-     endif()
-     message(STATUS
-       \"Installing: ${package_name} as flat Python egg under \${install_dir}\")
-     file(TO_NATIVE_PATH \"${CMAKE_INSTALL_PREFIX}\" install_prefix)
-     file(TO_NATIVE_PATH \"${CMAKE_INSTALL_PREFIX}/bin\" scripts_install_dir)
-     execute_process(
-       COMMAND
-         \"${PYTHON_EXECUTABLE}\" setup.py install
-           --single-version-externally-managed
-           --install-scripts \${scripts_install_dir}
-           --prefix \${install_prefix}
-           --record install.log
-           \${extra_install_args}
-       WORKING_DIRECTORY \"${build_dir}\"
-       OUTPUT_QUIET
-     )"
-  )
+    if(WIN32)
+      set(_ext "bat")
+    else()
+      set(_ext "sh")
+    endif()
+
+    configure_file(
+      "${ament_cmake_python_DIR}/environment_hooks/${hook_name}.${_ext}.in"
+      "${CMAKE_CURRENT_BINARY_DIR}/ament_cmake_python/environment_hooks/${hook_name}.${_ext}"
+      @ONLY
+    )
```

Comment:
I'm super confused by this, why is this only needed when sysmlink installing?

Reply:
This one's lovely. 

Whenever you do `python setup.py develop`, it lands an `.egg-link` file at the install prefix (i.e. a cross-platform symbolic link) and updates [a `.pth` file](https://docs.python.org/3/library/site.html) in the local (or global) `site-packages` directory with the content of that symbolic link. Alas, quoting a `pip` maintainer in https://github.com/pypa/pip/issues/435#issuecomment-3566380:

>  the .egg-link file is actually just a marker file; what causes the package to be importable is the entry that is added to easy-install.pth. pth files are a built-in feature of Python

When the interpreter is executed, it looks for these `.pth` files and appends their content to the `sys.path`.

Thing is, this logic does not play along with ROS overlays. As https://github.com/colcon/colcon-core/pull/163 explains, since the content of these `.pth` files is appended to the `sys.path`, it  hijacks the carefully crafted package order found in the PYTHONPATH. 

So what does `colcon` do? It adds [an environment hook](https://github.com/colcon/colcon-core/blob/83094da173d80559c92487fd1caca0a249e85b11/colcon_core/task/python/build.py#L139-L141) that prepends the PYTHONPATH with the path to the build directory where [package sources have been symlinked](https://github.com/colcon/colcon-core/blob/83094da173d80559c92487fd1caca0a249e85b11/colcon_core/task/python/build.py#L100). And voilà, the order is re-established, import lookups will effectively ignore `.pth` files' content (and thus we stopped generating them, see https://github.com/colcon/colcon-core/pull/219).

This patch does pretty much the same. It is not pretty, but because Python development installs are not pretty. I can see some on-going discussion in https://discuss.python.org/t/specification-of-editable-installation/1564 and https://discuss.python.org/t/standardising-editable-mode-installs-runtime-layout-not-hooks/4098. Perhaps it gets fixed in the future.

Reply:
:exploding_head: :exploding_head: :exploding_head: :exploding_head: 

ok, yeah not wonderful... but if it works and colcon was already doing the same I'm ok with it.